src/HexagonOffload.cpp

```diff
@@ -0,0 +1,322 @@
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    void visit(const Store *op) {
+        auto i = replacements.find(op->name);
+        if (i != replacements.end()) {
+            stmt = Store::make(op->name, mutate(op->value), mutate(op->index), i->second);
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+public:
+    ReplaceParams(const std::map<std::string, Parameter> &replacements)
+        : replacements(replacements) {}
+};
+
+Stmt replace_params(Stmt s, const std::map<std::string, Parameter> &replacements) {
+    return ReplaceParams(replacements).mutate(s);
+}
+
+
+class InjectHexagonRpc : public IRMutator {
+    std::map<std::string, Expr> state_vars;
+
+    Module device_code;
+
+    /** Alignment info for Int(32) variables in scope, so we don't
+     * lose the information when creating Hexagon kernels. */
+    Scope<ModulusRemainder> alignment_info;
+
+    Expr state_var(const std::string& name, Type type) {
+        Expr& var = state_vars[name];
+        if (!var.defined()) {
+            Buffer storage(type, {}, nullptr, name + "_buf");
+            *(void **)storage.host_ptr() = nullptr;
+            var = Load::make(type_of<void*>(), name + "_buf", 0, storage, Parameter());
+        }
+        return var;
+    }
+
+    Expr state_var_ptr(const std::string& name, Type type) {
+        Expr var = state_var(name, type);
+        return Call::make(Handle(), Call::address_of, {var}, Call::Intrinsic);
+    }
+
+    Expr module_state() {
+        return state_var("hexagon_module_state", type_of<void*>());
+    }
+
+    Expr module_state_ptr() {
+        return state_var_ptr("hexagon_module_state", type_of<void*>());
+    }
+
+    // Create a Buffer containing the given buffer/size, and return an
+    // expression for a pointer to the first element.
+    Expr buffer_ptr(const uint8_t* buffer, size_t size, const char* name) {
+        Buffer code(type_of<uint8_t>(), {(int)size}, nullptr, name);
+        memcpy(code.host_ptr(), buffer, (int)size);
+
+        Expr ptr_0 = Load::make(type_of<uint8_t>(), name, 0, code, Parameter());
+        return Call::make(Handle(), Call::address_of, {ptr_0}, Call::Intrinsic);
+    }
+
+    Stmt call_extern_and_assert(const std::string& name, const std::vector<Expr>& args) {
+        Expr call = Call::make(Int(32), name, args, Call::Extern);
+        string call_result_name = unique_name(name + "_result");
+        Expr call_result_var = Variable::make(Int(32), call_result_name);
+        return LetStmt::make(call_result_name, call,
+                             AssertStmt::make(EQ::make(call_result_var, 0), call_result_var));
+    }
+
+    using IRMutator::visit;
+
+    void visit(const For *loop) {
+        if (loop->device_api == DeviceAPI::Hexagon) {
+            // Unrolling or loop partitioning might generate multiple
+            // loops with the same name, so we need to unique them.
+            std::string hex_name = "hex_" + loop->name;
```

Comment:
Hmm, I think this comment might be stale. But, just for sanity's sake, I'll unique it anyways.